### PR TITLE
providers: delete instances via Executor object

### PR DESCRIPTION
**Pull request by mr-cal**
- [X] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [X] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `make lint`?
- [x] Have you successfully run `pytest tests/unit`?

-----

Previously, `snapcraft` bypassed `craft-providers` `Executor` object to delete instances. 
This low-level delete call causes issues when `craft-providers` modifies the name of the actual instance. (which happens [here](https://github.com/canonical/craft-providers/pull/134))

Instances are now deleted via the `Executor` object.

(CRAFT-1249)

### CODE REVIEW

File: snapcraft/providers/_lxd.py [hunk 1]
```diff
@@ -141,6 +124,24 @@ def is_provider_available(cls) -> bool:
         """
         return lxd.is_installed()
 
+    def environment(
+        self,
+        *,
+        instance_name: str,
+    ) -> Executor:
```
Comment (facundobatista) [inline snapcraft/providers/_lxd.py:131]: Why not all in the same line?

File: requirements.txt [hunk 2]
```diff
@@ -8,7 +8,7 @@ click==8.1.3
 craft-cli==1.1.0
 craft-grammar==1.1.1
 craft-parts==1.10.2
-craft-providers==1.3.1
+git+https://github.com/canonical/craft-providers#egg=craft-providers
```
Comment (facundobatista) [inline requirements.txt:11]: Let's release craft-providers, I need it too.

File: snapcraft/providers/_lxd.py (hunk 1 above)
Comment (mr-cal) [reply snapcraft/providers/_lxd.py]: Fixed

File: requirements.txt (hunk 2 above)
Comment (mr-cal) [reply requirements.txt]: `craft-providers` is released!

File: snapcraft/providers/_provider.py [hunk 3]
```diff
@@ -133,6 +148,15 @@ def is_provider_available(cls) -> bool:
         :returns: True if installed.
         """
 
+    @abstractmethod
+    def environment(self, *, instance_name: str) -> Executor:
```
Comment (sergiusens) [inline snapcraft/providers/_provider.py:152]:
```suggestion
    def create_environment(self, *, instance_name: str) -> Executor:
```

Comment (sergiusens) [reply snapcraft/providers/_provider.py:152]: That's my one comment for this PR, rest looks good.

Comment (mr-cal) [reply snapcraft/providers/_provider.py]: Renamed.

**Review by sergiusens — APPROVED** [review]

**Pull request merged**